subprojects/base-services/src/test/groovy/org/gradle/internal/hash/HashingTest.groovy

```diff
@@ -16,16 +16,81 @@
 
 package org.gradle.internal.hash
 
+
 import spock.lang.Specification
 
+import static java.lang.Thread.currentThread
+import static java.util.concurrent.CompletableFuture.supplyAsync
+import static java.util.concurrent.Executors.newFixedThreadPool
+
 class HashingTest extends Specification {
+    def 'cannot call hash multiple times'() {
+        given:
+        def hasher = Hashing.newHasher()
+        hasher.putInt(1)
+
+        when:
+        hasher.hash()
+        hasher.hash()
+
+        then:
+        thrown(IllegalStateException)
+    }
+
+    def 'nested hashers works'() {
+        when:
+        def hasher = Hashing.newHasher()
+        hasher.putInt(1)
+
+        and:
+        def nestedHasher = Hashing.newHasher().with(true) { it.putInt(1) }
+
+        then:
+        nestedHasher.hash() == hasher.hash()
+    }
+
+    def 'hasher works without calling final hash method'() {
+        given:
+        def value = ('a'..'z').join()
+
+        when:
+        def hasher1 = Hashing.newHasher()
+        hasher1.putString(value)
+        def hash = hasher1.hash()
+
+        and:
+        def hasher2 = Hashing.newHasher()
+        hasher2.putString(value)
+        // no call to hasher2.hash()
+
+        and:
+        def hasher3 = Hashing.newHasher()
+        hasher3.putString(value)
+        def hash3 = hasher3.hash()
+
+        then:
+        hash == hash3
+    }
```

Comment:
Please let me know if the above explanation isn't satisfactory.
This is an important test, will gladly explain, if needed.

Reply:
This test tests if there's a connection between different hashers. Why would that be the case?

Reply:
Because that was the case with the `ThreadLocals`. I've added these tests to document why we're not reusing the hashers.
Please let me know explicitly if it doesn't make sense to you, and I'll remove them.